### ISSUE TITLE
x64 Frida cmpcov/cmplog

### ISSUE THREAD

**Issue by expend20** [AFLplusplus/LibAFL#823]
> Is your feature request related to a problem? Please describe.

I'm always frustrated when fuzzer can't pass `cmp reg, 1337h`-like check :)

Okay, that's just work in progress thing, I'd like to keep open/public, so that everyone interested could join the discussion.

I'm about to implement cmpcov/cmplog for Frida for x64 architecture, and there are few possible scenarios which can be done. Let's first consider the only public approach which I'm aware of at the moment, which is [Jackalope/TinyInst/LiteCov](https://github.com/googleprojectzero/TinyInst/blob/master/arch/x86/x86_litecov.cpp#L194)

## TinyInst

The input assembly `cmp     dword ptr [rcx], 37333331h` is translated into:

```asm
nop5 ; space left for jump, when all 4 bytes passed once, the check is completely removed
push rax ; randomly selected register to load the value from memory is preserved
mov eax, [rcx] ; the compared value loaded into temp register
xor     eax,37333331h ; bit comparison with target value
lzcnt   eax,eax ; count the number of zero bits in the result
cmp     eax,8 ; compare the first byte, when the first byte is passed this check is patched to 16, then 24, then 32
jb @f ; skip logging the coverage
mov     byte ptr [cmpcov_buf],al ; logging the coverage
@@:
pop rax ; restoration of temp register
```

Few highlights:
* In case of reg/imm comparison there is no need to load value from the memory, so one instruction is saved
* the check is modified every time new coverage is reported, and cleared in the end leading to nearly zero overhead 

## Direct cmp

Another approach which I used once is more straight forward, `cmp     dword ptr [rcx], 37333331h` is translated into:

```asm
push rax ; preserve random register
mov eax, [rcx] ; load value from memory
cmp al, 31h ; first byte comparison
jnz @exit
mov [cmpcov_buf + 0], 1
shr eax, 8
cmp al, 33h ; second byte comparison
jnz @exit
mov [cmpcov_buf + 1], 1
shr eax, 8
cmp al, 33h ; third byte comparison
jnz @exit
mov [cmpcov_buf + 2], 1
shr eax, 8
cmp al, 37h ; fourth byte comparison
jnz @exit
mov [cmpcov_buf + 3], 1
@exit:
pop rax
```
Few highlights:
* In case of the register (not immediate) comparison, we need to preserve and shift two registers all the time
* Instead of jumping to the `@exit` all the time, we can jump to next check, that way we can match bytes in the middle, but it's more pricy (perfrormace cost). So, the performance is degraded, but the interesting samples could be discovered with less iterations.
* The check is placed only once and never modified (in contrast to Jackalope), this is much more overhead, but the feedback is more sensitive

## Call to the function

Alternatively we can just issue a call each time cmp-like instruction is met, and deal with two values in Rust code, that's how it's currently implemented in arm for Frida if I'm not mistaken. 

## Questions to think of

* Which approach is prefferrable?
* Do we need to unpatch the check once it's passed?
* Do we need to do cmplog or cmpcov is sufficient?
* What did I miss?

**Comment by s1341** [comment]
I think it will be best to follow the aarch64 implementation and do full cmplog. Then when redqueen is implemented in LibAFL mutator, we’ll get it for free. 

I think you will have a hard time patching and repatching the same basic blocks using frida stalker. You only get to rewrite once with stalker.

**Comment by s1341** [comment]
Also in the second example did you mean to shift right by 8 bits? Because you only shifted right by 4 bits each time.

**Comment by expend20** [comment]
> Also in the second example did you mean to shift right by 8 bits? Because you only shifted right by 4 bits each time.

sure, fixed, thanks

**Comment by andreafioraldi** [comment]
Counting 0s in the XOR is called value profile, and it is not a TinyInst thing, you find it already implemented in LibAFL for src level instrumentation at https://github.com/AFLplusplus/LibAFL/blob/main/libafl_targets/src/value_profile.h

**Comment by andreafioraldi** [comment]
While value profile can in theory bypass more checks than cmplog, practically it generates many intermediate testcases resulting in corpus bloating on many targets. That's why in libfuzzer this option is disabled by default.

**Comment by andreafioraldi** [comment]
> Which approach is prefferrable?

CmpLog is far better, AFL++ hongfuzz and libfuzzer all implement something similar by default

> Do we need to unpatch the check once it's passed?

It would be cool, but not essential, the performance gain are not massive for few checks and with a fuzzer that can linearly scale across core we can afford a small single core slowdown

> Do we need to do cmplog or cmpcov is sufficient?

We firstly need CmpLog (and we have it in libafl_frida), CompCov or value profile are a plus but nice to have

> What did I miss?

The real difference between CompCov (cmp byte-level splitting) and value profile (counting the zeros of the xor) is that both are generating intermediate testcases but value profile is bloating the corpus (it has more "sensitivity" is the correct term in literature) a bit more

**Comment by expend20** [comment]
thanks @andreafioraldi!

> We firstly need CmpLog (and we have it in libafl_frida)

that's the origin of this issue, x86 is missing, I'm actually only interested in x64

> both are generating intermediate testcases but value profile is bloating the corpus

In case new value passed, all previous corpus samples which only had partial value match is unnecessary anymore and could be removed

**Comment by andreafioraldi** [comment]
Oh yes i forgot it is arm only. AFL++ frida mode has it for x86 btw, i think you can almost copy paste the code.

For the remove from the corpus, it isn't really important as you can use only a subset of the corpus triggering the total coverage with the indexlenminimizer scheduler (like AFL does). Even doing it, we still observer a minimal benefit compared to CmpLog, checkout the libafl paper we measured them even in combination.

**Comment by expend20** [comment]
thanks, that's helpful

**Comment by expend20** [comment]
I've drafted the POC PR, the code seems correct, it extracts the arguments from the memory/registers, and then inserts a call to __libafl_targets_cmplog_instructions in the end. The only issue is that there is no actual new path discovery  
Maybe I misunderstand something, here is sample instruction `cmp dword ptr [rax],1337h`, which is translated into:
```
00000243`f9263652 50                   push    rax
00000243`f9263653 51                   push    rcx
00000243`f9263654 52                   push    rdx
00000243`f9263655 4150                 push    r8
00000243`f9263657 4151                 push    r9
00000243`f926365a 48b90000000000000000 mov     rcx,0               ; self pointer is not used 
00000243`f9263664 8b9000000000         mov     edx,dword ptr [rax] ; first cmp argument 
00000243`f926366a 49b83713000000000000 mov     r8,1337h            ; second cmp argument
00000243`f9263674 49b90000000000000000 mov     r9,0                ; ret address is temporarily nulled
00000243`f926367e ff1502000000         call    qword ptr [00000243`f9263686]
00000243`f9263684 eb08                 jmp     00000243`f926368e
...
00000243`f926368e 4159                 pop     r9
00000243`f9263690 4158                 pop     r8
00000243`f9263692 5a                   pop     rdx
00000243`f9263693 59                   pop     rcx
00000243`f9263694 58                   pop     rax
00000243`f9263695 813837130000         cmp     dword ptr [rax],1337h

0:000> dqs [00000243`f9263686]
00000243`f9263686  00007ff6`29245f20 frida_gdiplus!libafl_frida::cmplog_rt::CmpLogRuntime::populate_lists [Z:\git\LibAFL\libafl_frida\src\cmplog_rt.rs @ 184]

frida_gdiplus!libafl_frida::cmplog_rt::CmpLogRuntime::populate_lists -> __libafl_targets_cmplog_instructions
```

cc @tokatoka 
> can you check if the pattern 1337h is recorded?

I can confirm it lands into meta:
```
meta: Some(CmpValuesMetadata { list: [U64((1337, 1bf0000))] })
meta: Some(CmpValuesMetadata { list: [U64((1337, 34333231))] })
meta: Some(CmpValuesMetadata { list: [U64((1337, 61646162))] })
meta: Some(CmpValuesMetadata { list: [U64((1337, 62626162))] })
meta: Some(CmpValuesMetadata { list: [U64((1337, 62626262))] })
meta: Some(CmpValuesMetadata { list: [U64((1337, 62646162))] })
meta: Some(CmpValuesMetadata { list: [U64((1337, 62a7a7a7))] })
```
tracing is also enabled. However there seems no attempt from the fuzzer to put one of the values of the operands into the sample corpus. The number of unique variants is somewhat limited, I'd expect much more unique options. Maybe I hit some edge case?  The fuzzing code is just a:
```c++
extern "C" __declspec(dllexport) size_t
    LLVMFuzzerTestOneInput(const char *data, unsigned int len) {
  if (len >= 4 && *(int *)data == 0x1337) {
    int volatile *p = 0x0;
    *p = 0;
  }
  return 0;
}
```
Any ideas?

**Comment by tokatoka** [comment]
this is done
